### PR TITLE
chore(rfcs): Remap language infallible functions

### LINKED ISSUE #4409: Add support to "catch" errors in remap language
Currently, any function in the remap language is allowed to fail. This can happen for example if you use one of the `parse_*` functions, such as `parse_json`, but you pass in a value that can't be parsed. 

You can configure the `remap` transform to either drop an event for which the script fails, or keep the original one, but this applies to the entire script you used to remap the event, meaning in a 100 line script, if one function fails, the entire script fails.

This might be too coarse of a control, as some people might want to act on a single failing function and define what should happen in the error case.

There are multiple ways we can solve this, this issue is simply to make sure we track this situation and come up with a long-term solution.

### PR DESCRIPTION
**Pull request by JeanMertz**
Another Remap RFC to work towards reducing runtime failures.

👀 [Rendered version](https://github.com/timberio/vector/blob/ab202147b5e06ba05c1dce3f8b09f12fce9b5ac7/rfcs/2020-11-02-remap-infallible-functions.md)

closes #4409

### CODE REVIEW

File: rfcs/2020-11-02-remap-infallible-functions.md [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+Users need a way to handle such situations, without having to resort to complex
+solutions.
+
+### Infallible Function Variants
+
+The solution proposed by this RFC is to introduce the concept of _infallible
+function variants_.
+
+An infallible function variant is an exact copy of an existing fallible
+function, but it is guaranteed to never fails.
```
Comment (jszwedko) [inline rfcs/2020-11-02-remap-infallible-functions.md:72]: Just to clarify, I think we mean "guaranteed by the user to never fail"? It can still fail and will result in the script being aborted if it does.

File: rfcs/2020-11-02-remap-infallible-functions.md [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+
+## Outstanding Questions
+
+- Should we use "safe by default" or "fallible by default"?
+- Should assigning an undefined value result in no action, or removing any
+  potential existing value of the target path/variable?
+- How should if statements and logical operators handle undefined values?
+- Should we provide compile-time error for fallibly calling always-infallible
+  functions (e.g. `now!()`)?
+- Should we add support for (in)fallible path queries in this RFC?
```
Comment (jszwedko) [inline rfcs/2020-11-02-remap-infallible-functions.md:368]: I'd defer this and the manual abort function. They seem like they can be handled separate, and after, this RFC.

File: rfcs/2020-11-02-remap-infallible-functions.md [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+If `message` contains any data that `parse_json` can't handle, it will return
+`None`.
+
+How that `None` gets processed is up to the expression that processes the value.
+
+For example, in the case of the `Assignment` expression (which is what's called
+when doing `.message = …`), `None` values skip assigning the value to the target
+path, returning `None` itself.
+
+Given this, the outcome of the above example would change from "exception,
```
Comment (jszwedko) [inline rfcs/2020-11-02-remap-infallible-functions.md:127]: I actually find it confusing that `.message = parse_json(.message)` would not assign `.message` if parsing fails. I would expect it to receive the value of `null`. That is, I would not expect the assignment to be conditional on whether the function executes successfully.

File: rfcs/2020-11-02-remap-infallible-functions.md [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+```javascript
+if parse_json(.message) {
+    // parsing succeeded, do it again!
+    .message = parse_json(.message)
+}
+```
+
+This comes at a performance cost, as the message has to be parsed twice.
+
+An alternative is to consider assigning "nothing" to a target as erasing that
```
Comment (jszwedko) [inline rfcs/2020-11-02-remap-infallible-functions.md:186]: I think this is preferable as the first example is a pretty big gotcha.

Rather than erased, I would set `.foo` to `null` though. This still has the issue you mention about differentiating a failed parse from parsing the JSON string `null`, but I'll leave an extended comment with a suggested alteration to this proposal that would allow handling of it.

**Review by jszwedko — COMMENTED** [review]
I like where this is going. I wanted to suggest one substantive change though: returning error values (similar to Go and Lua). That is, the signature of `parse_json` would be `parse_json(string) -> value, error`.

I think this has two advantages:

* Being able to differentiate from a failed function call and a successful function call that returns `false` or `null`
* Being able to log the error message, if the user wants

Example:

```
if $message, $err = parse_json(.message); $err != null {
  // parsing failed
  log($err)
} {
  // parsing succeeded
}
```

We could allow users to not consume the error value if they don't care (without needing to provide the `_` that Go requires):

```
$message = parse_json(.message) // would ignore any error
```

I kind of like the requirement of using `_` though as it makes it clear that the error is being ignored. Otherwise, users will be less likely to realize that the function can fail.

File: rfcs/2020-11-02-remap-infallible-functions.md (hunk 3 above)
Comment (FungusHumungus) [reply rfcs/2020-11-02-remap-infallible-functions.md:127]: I agree. I think for the exact reason mentioned on line 168 that it should pass the `undefined` value on. (Note `null` is a valid value, so we do need the distinction between `null` and `undefined` here).

File: rfcs/2020-11-02-remap-infallible-functions.md [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+well onto the existing proposal with only minor modifications needed.
+
+### Infallible Or Fallible By Default
+
+The current RFC proposes to make infallible functions the default (e.g.
+functions without a bang will never fail). It requires an explicit decision by
+the user to add the `!` character and allow the function to fail.
+
+While the idea is that this is the most simple solution to the problem, there is
+a case to be made for the fact that this can unintentionally silently ignore
```
Comment (FungusHumungus) [inline rfcs/2020-11-02-remap-infallible-functions.md:280]: We should at least emit a warning, so it isn't completely silently ignored.

**Comment by JeanMertz** [comment]
@jszwedko 

> I think this has two advantages:
> 
>     * Being able to differentiate from a failed function call and a successful function call that returns `false` or `null`
>     * Being able to log the error message, if the user wants

I can't deny the advantages of this, but years of writing Go has made me put off by this pattern.

Aside from my personal feelings, I also think I don't really think this pattern fits great with the language as is. To me, it adds more cognitive complexity to the language than is worth doing.

Logging the original error message is definitely nice, but I don't think it carries its weight enough to justify the added complexity (and there are other, but perhaps less nice, ways to get the error of the last failed expression).

Maybe others feel differently, but to me, using something akin to `Result` type (as @Hoverbear has advocated for at one point) and the `value, err` pattern as you are advocating for are too heavy-handed to be a good fit for a simple language such as Remap.

The `!` vs non-`!` strikes a nice balance between giving more control to users and not over-complicating the language too much. There are other ways to do so, so I'm definitely open to discussing alternatives, though.

Having said all this, I'm going to add your proposal to the "alternatives" section, as it's obviously a pattern many people will be familiar with, and might even prefer 👍

File: rfcs/2020-11-02-remap-infallible-functions.md (hunk 1 above)
Comment (JeanMertz) [reply rfcs/2020-11-02-remap-infallible-functions.md:72]: Hmm, I'm not sure if we're on the same page here.

Calling a function as infallible (i.e. without appending the `!` to the call identifier), _will_ guarantee it never fails.

Comment (FungusHumungus) [reply rfcs/2020-11-02-remap-infallible-functions.md:72]: I think the terminology might be confusing. By 'fail' I think you mean it mill cause the whole script to abort with an error - or in the case of a condition resolve to false? An infallible one doesn't fail on an error, it returns 'undefined' and the script can choose how to handle this result.

**Comment by jszwedko** [comment]
> Logging the original error message is definitely nice, but I don't think it carries its weight enough to justify the added complexity (and there are other, but perhaps less nice, ways to get the error of the last failed expression).

Yeah, I could see also doing a "last error" variable like bash's `$?` if we don't want to make it part of the return signature, maybe `$ERR`? My preference is still for an explicit return, but I'll yield to the majority :smile: 

The fact that the assignment can result in assigning the value of "undefined" still feels a little weird to me as I might think of `=` _as_ defining it. I'm struggling to come up with a better term though. I also know that `null` vs. `undefined` sometimes confuses people in JS land (https://stackoverflow.com/questions/5076944/what-is-the-difference-between-null-and-undefined-in-javascript).

**Comment by FungusHumungus** [comment]
> The fact that the assignment can result in assigning the value of "undefined" still feels a little weird to me as I might think of `=` _as_ defining it. I'm struggling to come up with a better term though. I also know that `null` vs. `undefined` sometimes confuses people in JS land

It's an unfortunate side effect of `null` being a valid value. I wonder if instead of `undefined` we could have an error object instead? It could hold the error message. Calling it `Error` is clear as to what it's purpose is. The function `is_error` is more obvious than `is_undefined`..

**Comment by JeanMertz** [comment]
> It's an unfortunate side effect of `null` being a valid value. I wonder if instead of `undefined` we could have an error object instead? It could hold the error message. Calling it `Error` is clear as to what it's purpose is. The function `is_error` is more obvious than `is_undefined`..

Without having considered all implications too much, I do like this proposal...

It could potentially make things quite a bit easier to work with, since we could have expressions return `Result<Value, Error>` instead of `Result<Option<Value>, Error>`.

We would then introduce a special `Error` enum variant. Something like `Error::Infallible`. The `Infallible` expression would transform any `Error` variant into `Error::Infallible`. The runtime would run each expression, and if an error occurs, it would match on the error variant. Any non `Infallible` error aborts execution, while `Error::Infallible` will trigger a debug event, but otherwise continue executing the next expression.

The `is_error` function would return `true` if an expression returns an error (which can only be an infallible error, since any other error type would abort execution before `is_error` would run).

This would remove `undefined` from the language, but keeps `Value::Null`.

The only thing still missing would be error inspection. If we stick with the `!` approach proposed in this RFC, then I would like to leave error inspection out of this RFC (we can use constants, variables, or any other special way to access the last error), but obviously if we don't like the `!` approach and want to go with another approach such as Go's error reporting, then error inspection would still be part of the RFC.

One thing to consider is what to do when the last expression is an infallible function that fails:

```rust
parse_json("{ INVALID }")
```

What value (or error?) does this program resolve to?

Technically, we could do something like this:

```rust
impl Runtime {
    pub fn execute(&mut self, object: &mut impl Object, program: &Program) -> Result<Option<Value>> {
        let mut values = program
            .expressions
            .iter()
            .map(|expression| match expression.execute(&mut self.state, object) {
				Ok(value) => Ok(Some(value)),
				Err(Error::Infallible(_)) => Ok(None),
				Err(err) => Err(err),
			})
            .collect::<Result<Vec<Option<Value>>>>()?;

        Ok(values.pop().flatten())
    }
}
```

That is to say, even though individual expressions no longer return an option, it's still possible for the program as a whole to return "no value", _iff_ the last expression in the program returned an `Error::Infallible` error.

There is also still the matter of assigning infallible function values to a variable:

```rust
.foo = true
.foo = parse_json("{ INVALID }")
.foo // what does this resolve to?
```

line 2 could do one of three things:

1. keep the original value of `.foo`
2. erase the value of `.foo` (e.g. delete the field)
3. set the value to `null`

I guess option `2` causes the least amount of surprise?

**Comment by FungusHumungus** [comment]
> I guess option `2` causes the least amount of surprise?

I think so. But we would want to really encourage the user to do some error checking to handle it. The documentation would need to be very explicit that this could happen.

**Comment by binarylogic** [comment]
This RFC is where I share @Hoverbear's concerns (https://github.com/timberio/vector/pull/4839#pullrequestreview-522704268). And apologies if any of my interpretation is off, I don't fully understand the low-level implementation details so I likely have some blind spots.

That said, introducing failable function variants feels like it's getting away from the core premise of the language, which is to be an ultra simple, zero learning curve, domain-specific language. `!` variants, error handling, and the like, starts to feel daunting and I'm afraid that'll turn users away from Vector.

## Proposal

To go back to your example:

```
.message = parse_json(.message)
.timestamp = now()
```

If `.message` is not valid JSON:

1. Execution aborts.
2. Any changes to the event are reverted; the event is untouched.
3. A component-level option, like `drop_failed` (similar to [`drop_invalid'](https://vector.dev/docs/reference/transforms/json_parser/#drop_invalid)), determines if we keep or discard the event.

In other words, all functions are `!` variants. If a user wants to handle errors, then we can provide advanced error handling capabilities, like `try/catch` blocks.

And if we wanted to get ultra-explicit about functions that fail execution, we can just add `!` to the end of their name. For example, there is no `parse_json` function, it would be `parse_json!`.

---

@JeanMertz this might be one of those RFCs where a scheduled chat would help establish consensus.

**Comment by jszwedko** [comment]
What a wet blanket @binarylogic is :smile: 

I actually am coming around to that idea though, that failures should drop or dead-letter events. I think it'll likely be rare that events only occasionally fail, where I would see error handling being more useful.

I think we'll want a way to allow users to drop or dead letter events even if we introduce this error handling later so I guess I'm inclined to just add that functionality for now and defer introducing error handling until after the initial release of remap and we start collecting user feedback.

**Review by lukesteensen — COMMENTED** [review]
I think I'm landing roughly on the same side of the fence as @binarylogic and @Hoverbear here. Here's roughly how I'm thinking about it:

1. `remap` is not a general-purpose language, but one specialized for data mapping and reshaping. It's easy to think of functions here as procedural, but they don't need to be.
2. We already have a concept of failure in that there may be no data at a given query path. This already has relatively intuitive handling that matches what's expected in a declarative-ish mapper.
3. Functions like parsing JSON can be thought of as a slightly more complex query, not just asserting that there is data present at a given key, but also that it is valid JSON. If the assumptions built into that query are false, we can handle it the same way we handle queries for unknown keys.

File: rfcs/2020-11-02-remap-infallible-functions.md [hunk 6]
```diff
@@ -0,0 +1,371 @@
+# RFC … - 2020-11-02 - Remap Language Fallible and Infallible Functions
+
+By introducing a distinction between fallible and infallible functions, we can
```
Comment (lukesteensen) [inline rfcs/2020-11-02-remap-infallible-functions.md:3]: The terminology here is a bit confusing to me. I would assume that "fallible" means a function that is able to fail (e.g. JSON parsing), and "infallible" means a function that cannot fail (e.g. `now()`).

File: rfcs/2020-11-02-remap-infallible-functions.md [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+The infallible function must be as easy (or easier) to call than the fallible
+function, without any significant syntactic complexity.
+
+### To Bang Or Not To Bang
+
+The proposal is to introduce the concept of the bang (`!`) function identifier
+additive to determine as a user if a function should be allowed to fail or not.
+
+Function identifiers with a bang are considered to be fallible, while functions
+without are not:
```
Comment (lukesteensen) [inline rfcs/2020-11-02-remap-infallible-functions.md:83]: As someone with a ruby background, I find this a bit confusing. There is a convention in that community that functions named with a bang will raise an exception on failure, while non-bang functions return nil. This makes sense to me because the `!` bring attention to the fact that it may blow up. It also makes sense in the case of remap because it feels like a context in which we should be forgiving of imperfect data.

If we do go with the bang idea (I'm not yet convinced), then I'd recommend reversing the meaning.

File: rfcs/2020-11-02-remap-infallible-functions.md (hunk 3 above)
Comment (lukesteensen) [reply rfcs/2020-11-02-remap-infallible-functions.md:127]: I think we have to weigh this confusion vs adding a second meaning of `null` as "operation failed". If we don't differentiate between the two, operations that legitimately return `null` could trigger unexpected behavior.

File: rfcs/2020-11-02-remap-infallible-functions.md [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+```
+
+### False vs. "Nothing"
+
+One issue with this proposal is that `false` is a valid JSON value, and so the
+above check does not distinguish between false values or failed JSON parsing.
+
+One solution to this is to _not_ update `IfStatement`, but instead introduce a
+new `is_undefined` function, which returns `true` for `None` values, or else
+returns `false` (or inversely, we add a `is_defined` function):
```
Comment (lukesteensen) [inline rfcs/2020-11-02-remap-infallible-functions.md:158]: If we tweak this a bit, you could imagine an `ok(expr)` function that is essentially `expr != None`. If assignment returns the assigned value, you could have something like:

```
if ok($message = parse_json(.message)) {
    ...
} else {
    ...
}
```

Alternatively, we could have assignment return a boolean of whether or not a value was assigned and skip the extra function altogether for this case.

**Review by lukesteensen — COMMENTED** [review]
I actually think the bang feature and function variants are the less interesting part of this proposal. If we consider a default behavior where null/undefined/none silently fall through, the bang is just a shorthand for `if fallible(...) { ... } else { abort }`, and I agree that manual aborts are something we can hold off on for now.

A simpler alternative seems to be focusing on how null/undefined/none flow through things like assignment and conditionals. If we do that well, I think the case for aborts is relatively weak. For example, consider a conditional-only version:

```
if .message = parse_json(.message) {
    ...
}
.timestamp = now()
```

And a bang version:

```
.message = parse_json!(.message)
.timestamp = now()
```

In my opinion, the conditional version makes it much clearer what will happen in the failure case and forces users to put things within a special block if they want them to only happen conditionally. That seems like a big win for predictability.

File: rfcs/2020-11-02-remap-infallible-functions.md (hunk 7 above)
Comment (binarylogic) [reply rfcs/2020-11-02-remap-infallible-functions.md:83]: Did someone say Ruby? 😄

**Comment by JeanMertz** [comment]
A lot has happened since opening this RFC. While we didn't drive the RFC itself to conclusion, many valuable discussions and proof-of-concepts came from it.

I'm going to close this, as we've settled on an error-handling story as detailed in [this gist](https://gist.github.com/JeanMertz/f36e0aae6191231d1a37d81ba560047e).

We're working towards implementing those error-handling solutions, starting with #5876, #5830 and #5865.
